# Yandex API

## Задание
Существует API, которое умеет отвечать по трём URL: /countries, /cities и /populations. Клиентское приложение подсчитывает численность населения в Африке. Запросы друг от друга не зависят. Чтобы браузер пользователя не простаивал, клиентскому приложению важно уметь делать все три запроса одновременно. Реализацией API является функция getData(url, callback), которая принимает строку с URL запроса и функцию обратного вызова. В случае ошибки в callback первым аргументом будет передана строка ошибки, в случае успеха вторым аргументом будет передан ответ API.

Вам досталась реализация клиентского приложения, которое должно решать описанную выше задачу. Но в коде приложения есть ошибки, из-за которых фактический результат работы отличается от ожидаемого. Сам код [здесь](https://gist.github.com/verkholantsev/4d14ce053b009dac1225).

Как должно быть: приложение выводит в консоль суммарную популяцию в Африке. Как на самом деле: приложение не выводит в консоль ничего.

Задание
* Найдите ошибку в коде приложения, из-за которой реальный результат работы отличается от ожидаемого. Опишите, как эта ошибка могла возникнуть и как её избежать в будущем.
* Добавьте в приложение новую возможность — диалог с пользователем. Приложение спрашивает название страны или города, а затем показывает численность населения. Для диалога можно использовать window.prompt.

Результат пришлите в виде двух ссылок: на работающий пример и на исходный код на GitHub.

## Описание и исправление ошибок
Основная логическая ошибка заключалась в том, что все переданные в ```getData()``` функции обратного вызова используют один и тот же URL, который равен ```'/populations'```. Это происходит потому, что цикл ```for``` не создает новую область видимости. На каждой итерации переменная ```request``` будет иметь одно и то же значение.

Решить данную проблему можно, обернув callback в функцию (замыкание) со своей областью видимости переменных:
```
function doRequest(request) {
    return function (error, result) {
        . . .
    };
}
```

Во внешний цикл следует передавать не саму функцию, а функцию-обертку, на вход которой подается необходимый URL:
```
for (count = 0; count < 3; count++){
    getData(requests[count], doRequest(requests[count]));
}
```
Подобная ошибка могла возникнуть из-за непонимания принципов устройства пространства имен в языке JavaScript.

Чтобы в дальнейшем не совершать подобную ошибку, нужно код, в котором используются собственные переменные, оборачивать в замыкания, чтобы не было проблем с областями видимости.

Также в коде были найдены следующие синтаксические и менее серьезные ошибки:
* стрк. 22 – лишняя запятая при инициализации массива объектов;
*	стрк. 41 – неправильное использование Math.random() – пропущены скобки;
*	некоторые переменные не были объявлены (i, j);
*	желательно вынести объявление переменных за пределы цикла ```for```;
*	желательно давать переменным понятные имена, отражающие их суть, состоящие более чем из одной-двух букв (c, cc, p).

## Демонстрация
[Рабочий пример на GitHub Pages](http://oktava6.github.io/Yandex_API/)

Исходный код выше.
